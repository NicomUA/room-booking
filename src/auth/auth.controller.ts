import { Body, Controller, Post } from '@nestjs/common';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { JwtPayloadResponse } from './dto/payload.dto';

@Controller()
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('login')
  async login(
    @Body() { email, password }: LoginDto,
  ): Promise<JwtPayloadResponse> {
    return this.authService.login({ email, password });
  }
}
