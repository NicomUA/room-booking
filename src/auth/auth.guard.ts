import { ConfigurationService } from '@app/configuration';
import {
  CanActivate,
  ExecutionContext,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Request } from 'express';
import { JWTPayload } from './dto/payload.dto';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private jwtService: JwtService,
    private config: ConfigurationService,
  ) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest();
    const token = this.extractTokenFromHeader(request);
    const JWT_SECRET = this.config.get('JWT_SECRET');

    if (!token) {
      throw new UnauthorizedException('Token is missing');
    }
    try {
      const payload = await this.jwtService.verifyAsync<JWTPayload>(token, {
        secret: JWT_SECRET,
      });
      request['user'] = { id: payload.sub, ...payload };
    } catch {
      throw new UnauthorizedException();
    }
    return true;
  }

  private extractTokenFromHeader(request: Request): string | undefined {
    const [type, token] = request.headers.authorization?.split(' ') ?? [];
    return type === 'Bearer' ? token : undefined;
  }
}
