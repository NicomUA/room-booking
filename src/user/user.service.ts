import { DbService } from '@app/db';
import { Injectable } from '@nestjs/common';

@Injectable()
export class UserService {
  constructor(private db: DbService) {}

  findAll() {
    return this.db.user.findMany({
      select: { id: true, name: true, email: true },
    });
  }

  findByEmail(email: string) {
    return this.db.user.findUnique({
      where: { email },
    });
  }

  findById(id: number) {
    return this.db.user.findUnique({
      select: { id: true, name: true, email: true },
      where: { id },
    });
  }
}
