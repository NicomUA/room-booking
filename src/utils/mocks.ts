export const userMock = {
  id: 1,
  name: 'test',
  email: 'test@test.com',
};

export const roomMock = {
  id: 1,
  name: 'test',
};

export const bookingMock = {
  id: 1,
  roomId: 1,
  userId: 1,
  user: userMock,
  startTime: new Date(),
  endTime: new Date(),
  createdAt: new Date(),
  updatedAt: new Date(),
};
