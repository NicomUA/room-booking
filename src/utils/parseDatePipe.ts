import { PipeTransform, Injectable, BadRequestException } from '@nestjs/common';

@Injectable()
export class ParseDatePipe implements PipeTransform {
  transform(value: any) {
    if (!value) {
      throw new BadRequestException('Date should be provided');
    }
    // ISO Date regular expression check
    if (!/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z/.test(value)) {
      throw new BadRequestException(
        'Invalid date format. Expected YYYY-MM-DDTHH:MM:SS.MMMZ',
      );
    }

    return new Date(value);
  }
}
