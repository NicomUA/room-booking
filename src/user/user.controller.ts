import {
  Controller,
  Get,
  Param,
  ParseIntPipe,
  UseGuards,
  Request,
} from '@nestjs/common';
import { UserService } from './user.service';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { User } from '@prisma/client';
import { AuthGuard } from '../auth/auth.guard';

@ApiTags('user')
@ApiBearerAuth()
@UseGuards(AuthGuard)
@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @Get()
  async findAll(): Promise<Partial<User>[]> {
    return this.userService.findAll();
  }

  @Get('me')
  async getMe(@Request() req): Promise<Partial<User>> {
    const { sub: id } = req.user;
    return this.userService.findById(id);
  }

  @Get(':id')
  async findById(
    @Param('id', ParseIntPipe) id: number,
  ): Promise<Partial<User>> {
    return this.userService.findById(id);
  }
}
