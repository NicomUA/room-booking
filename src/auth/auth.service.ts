import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { LoginDto } from './dto/login.dto';
import { JwtPayloadResponse } from './dto/payload.dto';
import { UserService } from '../user/user.service';

@Injectable()
export class AuthService {
  constructor(
    private jwtService: JwtService,
    private userService: UserService,
  ) {}

  async login({ email, password }: LoginDto): Promise<JwtPayloadResponse> {
    const user = await this.userService.findByEmail(email);

    if (!user) {
      throw new NotFoundException('User not found');
    }

    if (user.password !== password) {
      throw new BadRequestException('Wrong password');
    }

    const payload = { sub: user.id, email: user.email, name: user.name };
    return {
      access_token: await this.jwtService.signAsync(payload),
    };
  }
}
