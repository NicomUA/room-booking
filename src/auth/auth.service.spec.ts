import { AuthService } from './auth.service';
import { TestBed } from '@automock/jest';
import { UserService } from '../user/user.service';
import { userMock } from '../utils/mocks';
import { JwtService } from '@nestjs/jwt';

describe('AuthService', () => {
  let service: AuthService;
  let userService: jest.Mocked<UserService>;
  let jwtService: jest.Mocked<JwtService>;

  beforeAll(async () => {
    const { unit, unitRef } = TestBed.create(AuthService).compile();
    service = unit;
    userService = unitRef.get(UserService);
    jwtService = unitRef.get(JwtService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('login', () => {
    it('should login', async () => {
      userService.findByEmail.mockResolvedValue({
        ...userMock,
        password: 'test',
      });
      jwtService.signAsync.mockResolvedValue('test');
      const result = await service.login({ email: 'test', password: 'test' });
      expect(result).toEqual({ access_token: 'test' });
    });

    it('should throw an error if user not found', async () => {
      userService.findByEmail.mockResolvedValue(null);
      await expect(
        service.login({ email: 'test', password: 'test' }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw an error if password is wrong', async () => {
      userService.findByEmail.mockResolvedValue({
        ...userMock,
        password: 'test',
      });
      await expect(
        service.login({ email: 'test', password: 'wrong' }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
