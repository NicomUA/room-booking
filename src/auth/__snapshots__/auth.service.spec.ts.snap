// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService login should throw an error if password is wrong 1`] = `"Wrong password"`;

exports[`AuthService login should throw an error if user not found 1`] = `"User not found"`;
